arangod/RocksDBEngine/RocksDBIndex.cpp

```diff
@@ -81,68 +66,49 @@ RocksDBIndex::RocksDBIndex(
     RocksDBEngine& engine)
     : Index(id, collection, name, attributes, unique, sparse),
       _cf(cf),
-      _cache(nullptr),
-      _cacheEnabled(useCache && !collection.system() &&
-                    cacheManager != nullptr),
-      _cacheManager(cacheManager),
+      _cacheManager(cacheManager != nullptr && !collection.system() &&
+                            !collection.isAStub() &&
+                            !ServerState::instance()->isCoordinator()
+                        ? cacheManager
+                        : nullptr),
+      _cacheEnabled(useCache && (_cacheManager != nullptr)),
       _engine(engine),
       _objectId(::ensureObjectId(objectId)) {
   TRI_ASSERT(cf != nullptr &&
              cf != RocksDBColumnFamilyManager::get(
                        RocksDBColumnFamilyManager::Family::Definitions));
-
-  if (_cacheEnabled) {
-    createCache();
-  }
-
-  _engine.addIndexMapping(_objectId.load(), collection.vocbase().id(),
-                          collection.id(), _iid);
+  _engine.addIndexMapping(_objectId, collection.vocbase().id(), collection.id(),
+                          _iid);
 }
 
 RocksDBIndex::RocksDBIndex(IndexId id, LogicalCollection& collection,
                            arangodb::velocypack::Slice info,
-                           rocksdb::ColumnFamilyHandle* cf, bool useCache)
-    : RocksDBIndex(id, collection, info, cf, useCache,
-                   collection.vocbase()
-                       .server()
-                       .getFeature<CacheManagerFeature>()
-                       .manager(),
-                   collection.vocbase()
-                       .server()
-                       .getFeature<EngineSelectorFeature>()
-                       .engine<RocksDBEngine>()) {}
-
-RocksDBIndex::RocksDBIndex(IndexId id, LogicalCollection& collection,
-                           arangodb::velocypack::Slice const& info,
                            rocksdb::ColumnFamilyHandle* cf, bool useCache,
                            cache::Manager* cacheManager, RocksDBEngine& engine)
     : Index(id, collection, info),
       _cf(cf),
-      _cache(nullptr),
-      _cacheEnabled(useCache && !collection.system() &&
-                    cacheManager != nullptr),
-      _cacheManager(cacheManager),
+      _cacheManager(cacheManager != nullptr && !collection.system() &&
+                            !collection.isAStub() &&
+                            !ServerState::instance()->isCoordinator()
+                        ? cacheManager
+                        : nullptr),
+      _cacheEnabled(useCache && (_cacheManager != nullptr)),
       _engine(engine),
       _objectId(::ensureObjectId(basics::VelocyPackHelper::stringUInt64(
           info, StaticStrings::ObjectId))) {
   TRI_ASSERT(cf != nullptr &&
              cf != RocksDBColumnFamilyManager::get(
                        RocksDBColumnFamilyManager::Family::Definitions));
-
-  if (_cacheEnabled) {
-    createCache();
```

Comment:
I made `createCache()` a virtual function, so that the RocksDBEdgeIndex and RocksDBVPackIndex can have different implementation.
Because RocksDBIndex is a base class of both and should not call virtual functions in its ctor, I moved the call to `createCache()` from the base class ctor to the ctors of each derived class.

Reply:
OK. Understood.